functorch/csrc/BatchRulesNorm.cpp

```diff
@@ -5,107 +5,292 @@
-    bool training, double momentum, double eps, bool cudnn_enabled) {
-  TORCH_INTERNAL_ASSERT(input_bdim.has_value());
-  auto batch_size = input.size(*input_bdim);
-  auto input_ = reshape_dim_into(*input_bdim, /*channels dim*/1, input);
-
-  auto weight_ = maybe_flatten(weight, weight_bdim);
-  auto bias_ = maybe_flatten(bias, bias_bdim);
-  auto running_mean_ = maybe_flatten(running_mean, running_mean_bdim);
-  auto running_var_ = maybe_flatten(running_var, running_var_bdim);
-
-  auto result = at::batch_norm(
-      input_, weight_, bias_, running_mean_, running_var_,
-      training, momentum, eps, cudnn_enabled);
-  return std::make_tuple(reshape_dim_outof(1, batch_size, result), 1);
+    const c10::optional<Tensor>& weight_opt, optional<int64_t> weight_bdim,
+    const c10::optional<Tensor>& bias_opt, optional<int64_t> bias_bdim,
+    const c10::optional<Tensor>& running_mean_opt, optional<int64_t> running_mean_bdim,
+    const c10::optional<Tensor>& running_var_opt, optional<int64_t> running_var_bdim,
+    bool training, double momentum, double eps) {
+  c10::MaybeOwned<Tensor> weight_maybe_owned = at::borrow_from_optional_tensor(weight_opt);
+  const Tensor& weight = *weight_maybe_owned;
+  c10::MaybeOwned<Tensor> bias_maybe_owned = at::borrow_from_optional_tensor(bias_opt);
+  const Tensor& bias = *bias_maybe_owned;
+  c10::MaybeOwned<Tensor> running_mean_maybe_owned = at::borrow_from_optional_tensor(running_mean_opt);
+  auto running_mean_ = *running_mean_maybe_owned;
+  c10::MaybeOwned<Tensor> running_var_maybe_owned = at::borrow_from_optional_tensor(running_var_opt);
+  auto running_var_ = *running_var_maybe_owned;
+
+  if (input_bdim && ((running_mean_.defined() && !running_mean_bdim) || (running_var_.defined() && !running_var_bdim))) {
+    throw std::runtime_error("Batch norm got a batched tensor as input while the running_mean or running_var, which will be updated in place, were not batched.");
+  }
+  auto input_ = input;
+  c10::optional<int64_t> bdim_size;
+  if (input_bdim) {
+    bdim_size = input.size(*input_bdim);
+    input_ = reshape_dim_into(*input_bdim, /*channels dim*/1, input);
+  } else if (running_mean_bdim) {
+    bdim_size = running_mean_.size(*running_mean_bdim);
+    input_ = repeatForUpdatedValues(input_, *bdim_size);
+  } else if (running_var_bdim) {
+    bdim_size = running_var_.size(*running_var_bdim);
+    input_ = repeatForUpdatedValues(input_, *bdim_size);
+  }
+
+  if (running_mean_bdim) {
+    running_mean_ = reshape_dim_into(*running_mean_bdim, 0, running_mean_);
+  }
+  if (running_var_bdim) {
+    running_var_ = reshape_dim_into(*running_var_bdim, 0, running_var_);
+  }
+
+  const auto dummy_weight = at::ones(input_.size(1), input_.options());
+  const auto dummy_bias = at::zeros(input_.size(1), input_.options()); // can't use efficientzeros because cudnn expects contiguous
+  const auto result = Func(input_, dummy_weight, dummy_bias, running_mean_, running_var_, training, momentum, eps);
```

Comment:
cudnn and miopen both require a weight tensor by the signature and they'll error that they need strides if you give them an empty tensor

It looks like they also require bias because I get the same strides error message if I don't pass it in so I updated the comment to reflect that

Reply:
What happens if we call F.batch_norm without a weight and bias using a CUDA tensor input? Does it go the native_batch_norm route or does it hit cudnn?

Reply:
It will route to native_batch_norm (there's a cuda implementation of native batch norm that I haven't looked at much)